go/proxyd/daisychain.go

```diff
@@ -0,0 +1,485 @@
+	chainIds := []*hexutil.Big{}
+	for _, backend := range s.Backends() {
+		res, _ := backend.Forward(context.Background(), req)
+		str, ok := res.Result.(string)
+		if !ok {
+			return nil, errors.New("cannot fetch chainid on start")
+		}
+		chainId := new(hexutil.Big)
+		err := chainId.UnmarshalText([]byte(str))
+		if err != nil {
+			return nil, err
+		}
+		chainIds = append(chainIds, chainId)
+	}
+
+	if len(chainIds) == 0 {
+		panic("cannot fetch remote chain id")
+	}
+	chainId := chainIds[0].ToInt()
+	for _, id := range chainIds {
+		if id.ToInt().Cmp(chainId) != 0 {
+			log.Crit("mismatched chain ids detected", "chain-id", chainId, "other", id)
+		}
+	}
+	log.Info("detected chain id", "value", chainId)
+	if s.l2ChainID != nil {
+		if s.l2ChainID.Cmp(chainId) != 0 {
+			return nil, fmt.Errorf("mismatched chainids: expected %d, got %d", s.l2ChainID, chainId)
+		}
+	}
+	return chainId, nil
+}
+
+// HandleRPC is the top level RPC handler for the daisychain.
+// All RPC requests are routed through this handler.
+func (s *DaisyChainServer) HandleRPC(w http.ResponseWriter, r *http.Request) {
+	ctx := populateContext(w, r, s.authenticatedPaths)
+	if ctx == nil {
+		return
+	}
+
+	doRequest := func(ctx context.Context, req *RPCReq) (*RPCRes, bool) {
+		argType, ok := argTypes[req.Method]
+		if !ok {
+			return NewRPCErrorRes(req.ID, ErrParseErr), false
+		}
+
+		values, err := parsePositionalArguments(req.Params, argType)
+		if err != nil {
+			return NewRPCErrorRes(req.ID, fmt.Errorf("%s: %w", ErrParseErr, err)), false
+		}
+
+		argument, ok := parseRequestOptions(values)
+		if !ok {
+			return NewRPCErrorRes(req.ID, ErrParseErr), false
+		}
+
+		req, err = trimRequestOptions(req, values)
+		if err != nil {
+			return NewRPCErrorRes(req.ID, ErrParseErr), false
+		}
+
+		// TODO(tynes): perhaps a better approach is to attempt to classify the request
+		// up front and then have an enum in a switch statement
+		var res *RPCRes
+		// Check to see if the request is meant for the latest node
+		// first. This will be the most commonly used case, so check first.
+		if s.isLatestEpochsRPC(argument) {
+			res = s.handleLatestEpochsRPC(ctx, req, values)
+		} else if s.isHashBasedRPC(values) {
+			// Check to see if a hash was passed in the rpc params
+			// and fall back to each node, looking for the hash
+			// TODO: If we want users to be able to pass a hash
+			// and not need to be explicit with the epoch and have
+			// it "just work" then this needs to be checked first
+			res = s.handleHashTaggedRPC(ctx, req)
+		} else {
+			// If an epoch was passed then route based on the epoch
+			res = s.handleEpochRPC(ctx, req, argument)
+		}
+		return res, false
+	}
+
+	handleRPC(ctx, w, r, s.maxBodySize, doRequest)
+}
+
+// isLatestEpochsRPC will check to see if the request
+func (s *DaisyChainServer) isLatestEpochsRPC(opts *RequestOptions) bool {
+	if opts == nil {
+		return true
+	}
+	if opts.Epoch == 5 || opts.Epoch == 6 {
```

Comment:
Why is Epoch 5 also considered the latest?

Reply:
Epoch 5 and epoch 6 are considered the latest because epoch 6 starts at a non zero block number. This means that when we upgrade to bedrock, the first block will be the height of the current system + 1